File: Data_Structure/Non_Linear/Binary_Tree/C++/BinaryTree.cpp
#include <bits/stdc++.h>
using namespace std;
template <class T>
class BinaryTree
{
private:
    struct Node
    {
        T val;
        Node *left = NULL;
        Node *right = NULL;
        Node(T val)
        {
            this->val = val;
        }
    };
	Node *root;
	public:
    BinaryTree() {         // empty constructor
		root = NULL;
	}; 
    void insert(T val)
    {
        Node *node = new Node(val);
        if (root == NULL)
        {
            root = node;
            return;
        }
        queue<Node *> q;
        q.push(root);
        while (!q.empty())
        {
            Node *temp = q.front();
            q.pop();
            if (temp->left == NULL)
            {
                temp->left = node;
                return;
            }
            else
            {
                q.push(temp->left);
            }
            if (temp->right == NULL)
            {
                temp->right = node;
                return;
            }
            else
            {
                q.push(temp->right);
            }
        }
    }
    void preOrderPrint(Node *node)
    {
        if (node == NULL)
        {
            return;
        }
        cout << node->val << " ";
        preOrderPrint(node->left);
        preOrderPrint(node->right);
    }
    void inOrderPrint(Node *node)
    {
        if (node == NULL)
        {
            return;
        }
        inOrderPrint(node->left);
        cout << node->val << " ";
        inOrderPrint(node->right);
    }
    void postOrderPrint(Node *node)
    {
        if (node == NULL)
        {
            return;
        }
        postOrderPrint(node->left);
        postOrderPrint(node->right);
        cout << node->val << " ";
    }
    void preOrderPrint()
    {
        preOrderPrint(root);
    }
    void inOrderPrint()
    {
        inOrderPrint(root);
    }
    void postOrderPrint()
    {
        postOrderPrint(root);
    }
    void deleteTree(Node *node)
    {
        if (node == NULL)
        {
            return;
        }
        deleteTree(node->left);
        deleteTree(node->right);
        delete node;
    }
    void BFS(Node* node){
        if(node == NULL){
            return;
        }
		queue<Node*> q;
		q.push(node);
		while (!q.empty())
		{
			Node* curr = q.front();

			if(curr->left != NULL){
				q.push(curr->left);
			}

			if(curr->right != NULL){
				q.push(curr->right);
			}
			
			cout<<curr->val<<" ";
			q.pop();
		}
    }
	void bfs(){
		BFS(root);
	}
    ~BinaryTree()
    {
        deleteTree(root);
    }
};
int main()
{
    
    BinaryTree<int> bt;
    bt.insert(1);
    bt.insert(2);
    bt.insert(3);
    bt.insert(4);
    bt.insert(5);
    bt.insert(6);
    bt.insert(7);
	cout<<"\n----------------------dfs------------------------\n";
    bt.preOrderPrint();
    cout << endl;
    bt.inOrderPrint();
    cout << endl;
    bt.postOrderPrint();
	cout<<"\n----------------------bfs------------------------\n";
	bt.bfs();
    
    return 0;
}

File: Data_Structure/Non_Linear/Binary_Tree/Java/BinaryTree.java
import java.util.LinkedList;
import java.util.Queue;

public class BinaryTree<T> {
    private Node<T> root;
    public BinaryTree() {
        this.root = null;
    }
    public void insert(T val) {
        if (this.root == null) {
            this.root = new Node<T>(val);
            return;
        }
        Queue<Node<T>> q = new LinkedList<Node<T>>();
        q.add(this.root);
        while (!q.isEmpty()) {
            Node<T> temp = q.remove();
            if (temp.left == null) {
                temp.left = new Node<T>(val);
                return;
            } else {
                q.add(temp.left);
            }
            if (temp.right == null) {
                temp.right = new Node<T>(val);
                return;
            } else {
                q.add(temp.right);
            }
        }
    }
    public void preOrderPrint(Node<T> node) {
        if(node == null){
            return;
        }
        System.out.print(node.val + " ");
        preOrderPrint(node.left);
        preOrderPrint(node.right);
    }
    public void inOrderPrint(Node<T> node) {
        if(node == null){
            return;
        }
        inOrderPrint(node.left);
        System.out.print(node.val + " ");
        inOrderPrint(node.right);
    }
    public void postOrderPrint(Node<T> node) {
        if(node == null){
            return;
        }
        postOrderPrint(node.left);
        postOrderPrint(node.right);
        System.out.print(node.val + " ");
    }
    public void deleteTree(Node<T> node) {
        if (node == null) {
            return;
        }
        deleteTree(node.left);
        deleteTree(node.right);
        node = null;
    }
    public void preOrderPrint(){
        preOrderPrint(this.root);
    }
    public void inOrderPrint(){
        inOrderPrint(this.root);
    }
    public void postOrderPrint(){
        postOrderPrint(this.root);
    }
	public void BFS(Node<T> node){
		if(node == null){
			return;
		}
		Queue<Node<T>> queue = new LinkedList<>();
		queue.add(node);
		while(!queue.isEmpty()){
			Node<T> curr = queue.remove();
			System.out.print(curr.val + " ");
			if(curr.left != null){
				queue.add(curr.left);
			}
			if(curr.right != null){
				queue.add(curr.right);
			}
		}
	}
	public void bfs(){
		BFS(root);
	}
}